File: backend/models/tech_platform.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column

from backend.models.base import Base
from backend.models.mixins import TimestampMixin, UUIDPrimaryKeyMixin


class TechPlatformDeploymentModule(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "tech_platform_deployment_modules"
    __table_args__ = (
        UniqueConstraint(
            "site_id", "dockerfile_path", name="uq_tech_platform_module_site_dockerfile"
        ),
    )

    project_id: Mapped[str] = mapped_column(
        ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True
    )
    site_id: Mapped[str] = mapped_column(
        ForeignKey("sites.id", ondelete="CASCADE"), nullable=False, index=True
    )
    dockerfile_path: Mapped[str] = mapped_column(String(512), nullable=False)
    build_context: Mapped[str] = mapped_column(String(512), nullable=False, default=".")
    app_name: Mapped[str] = mapped_column(String(255), nullable=False)
    namespace: Mapped[str] = mapped_column(
        String(255), nullable=False, default="ocean-km"
    )
    harbor_project: Mapped[str] = mapped_column(
        String(255), nullable=False, default="ocean-km"
    )
    repository_name: Mapped[str] = mapped_column(String(255), nullable=False)
    app_type: Mapped[str] = mapped_column(String(16), nullable=False, default="2")
    container_port: Mapped[int] = mapped_column(Integer, nullable=False, default=8080)
    service_port: Mapped[int] = mapped_column(Integer, nullable=False, default=80)
    config_map_template: Mapped[str] = mapped_column(Text, nullable=False, default="")
    deployment_template: Mapped[str] = mapped_column(Text, nullable=False, default="")
    service_template: Mapped[str] = mapped_column(Text, nullable=False, default="")
    platform_app_id: Mapped[str] = mapped_column(
        String(128), nullable=False, default=""
    )
    is_available: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    last_task_id: Mapped[str | None] = mapped_column(
        ForeignKey("agent_tasks.id", ondelete="SET NULL"), nullable=True, default=None
    )
    last_image: Mapped[str] = mapped_column(String(1024), nullable=False, default="")
    last_commit_sha: Mapped[str] = mapped_column(String(64), nullable=False, default="")
    status: Mapped[str] = mapped_column(String(32), nullable=False, default="idle")
    last_error: Mapped[str] = mapped_column(Text, nullable=False, default="")
    last_deployed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True, default=None
    )

File: frontend/src/api/techPlatform.ts
import client from './client'

export type TechPlatformModuleStatus = 'idle' | 'queued' | 'running' | 'success' | 'failed'

export interface TechPlatformDeploymentModule {
  id: string
  project_id: string
  site_id: string
  site_name: string
  dockerfile_path: string
  build_context: string
  app_name: string
  namespace: string
  harbor_project: string
  repository_name: string
  app_type: string
  container_port: number
  service_port: number
  config_map_template: string
  deployment_template: string
  service_template: string
  platform_app_id: string
  is_available: boolean
  last_task_id: string
  last_image: string
  last_commit_sha: string
  status: TechPlatformModuleStatus
  last_error: string
  last_deployed_at: string | null
  created_at: string | null
  updated_at: string | null
}

export type TechPlatformModuleInput = Pick<
  TechPlatformDeploymentModule,
  | 'site_id'
  | 'dockerfile_path'
  | 'build_context'
  | 'app_name'
  | 'namespace'
  | 'harbor_project'
  | 'repository_name'
  | 'app_type'
  | 'container_port'
  | 'service_port'
  | 'config_map_template'
  | 'deployment_template'
  | 'service_template'
>

export interface RenderedYamlResource {
  kind: 'ConfigMap' | 'Deployment' | 'Service'
  yaml: string
}

export const techPlatformAPI = {
  list(projectId: string) {
    return client.get<any, { ok: boolean; modules: TechPlatformDeploymentModule[] }>(
      `/projects/${projectId}/tech-platform/modules`,
    )
  },

  scan(projectId: string) {
    return client.post<any, { ok: boolean; modules: TechPlatformDeploymentModule[] }>(
      `/projects/${projectId}/tech-platform/modules/scan`,
    )
  },

  create(projectId: string, data: TechPlatformModuleInput) {
    return client.post<any, { ok: boolean; module: TechPlatformDeploymentModule }>(
      `/projects/${projectId}/tech-platform/modules`,
      data,
    )
  },

  update(projectId: string, moduleId: string, data: Partial<TechPlatformModuleInput>) {
    return client.patch<any, { ok: boolean; module: TechPlatformDeploymentModule }>(
      `/projects/${projectId}/tech-platform/modules/${moduleId}`,
      data,
    )
  },

  remove(projectId: string, moduleId: string) {
    return client.delete(`/projects/${projectId}/tech-platform/modules/${moduleId}`)
  },

  preview(projectId: string, moduleId: string, image = '') {
    return client.post<any, { ok: boolean; image: string; resources: RenderedYamlResource[] }>(
      `/projects/${projectId}/tech-platform/modules/${moduleId}/preview`,
      { image: image || null },
    )
  },

  validate(projectId: string, moduleId: string, image = '') {
    return client.post<any, { ok: boolean; valid: boolean; resources: RenderedYamlResource[] }>(
      `/projects/${projectId}/tech-platform/modules/${moduleId}/validate`,
      { image: image || null },
    )
  },

  deploy(projectId: string, moduleId: string) {
    return client.post<any, { ok: boolean; task_id: string }>(
      `/projects/${projectId}/tech-platform/modules/${moduleId}/deploy`,
    )
  },
}
